### PR TITLE
 Pass around i1 as is instead of going through u8

### PR DESCRIPTION
**Pull request by nox**

### CODE REVIEW

**Comment by nox** [comment]
Cc @rkruppe 
r? @eddyb
@bors-servo try

File: src/librustc_trans/mir/rvalue.rs [hunk 1]
```diff
@@ -114,7 +114,9 @@ impl<'a, 'tcx> FunctionCx<'a, 'tcx> {
                     }
 
                     // Use llvm.memset.p0i8.* to initialize byte arrays
-                    let v = base::from_immediate(&bx, v);
+                    if common::val_ty(v) == Type::i1(bx.cx) {
+                        v = bx.zext(v, Type::i8(bx.cx))
+                    }
```
Comment (nox) [inline src/librustc_trans/mir/rvalue.rs]: Am not sure about this, but I feel like boolean arrays should also use `memset` when they can, right?

Comment (eddyb) [reply src/librustc_trans/mir/rvalue.rs]: Indeed.

**Comment by eddyb** [comment]
cc @Mark-Simulacrum We're curious about a perf run here.

**Comment by Mark-Simulacrum** [comment]
Perf queued.

**Comment by nox** [comment]
I can't reproduce that locally, I wonder if this is an intermittent, or a failure specific to using LLVM 3.9.

**Comment by nox** [comment]
Seems like this was indeed an intermittent.

File: src/librustc_trans/type_of.rs [hunk 2]
```diff
@@ -359,7 +352,7 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyLayout<'tcx> {
         };
         let scalar = [a, b][index];
 
-        // Make sure to return the same type `immediate_llvm_type` would,
+        // Make sure to return the same type `llvm_type` would,
```
Comment (eddyb) [inline src/librustc_trans/type_of.rs]: Hmm, looking at this again, the special-case for `bool` could go better in `scalar_llvm_type_at`, which would let you remove this whole comment too.

**Comment by pietroalbini** [comment]
Ping from triage @nox! It's been a while since we heard from you, will you have time to work on this again soon?

**Comment by pietroalbini** [comment]
Ping from triage @nox! It's been a while since we heard from you, will you have time to work on this again soon?

**Pull request closed without merging**